quickwit-swim/src/membership.rs

```diff
@@ -133,55 +125,74 @@ impl ArtilleryMemberList {
         None
     }
 
+    fn find_member<'a>(
+        members: &'a mut [ArtilleryMember],
+        node_id: &str,
+        addr: SocketAddr,
+    ) -> Option<&'a mut ArtilleryMember> {
+        // ugly hack for borrowchecker
+        if members.iter().any(|member| member.node_id() == node_id) {
+            members
+                .iter_mut()
+                .find(|member| member.node_id() == node_id)
+        } else {
+            members
+                .iter_mut()
+                .find(|member| member.remote_host() == Some(addr))
+        }
+    }
+
     pub fn apply_state_changes(
         &mut self,
         state_changes: Vec<ArtilleryStateChange>,
         from: &SocketAddr,
     ) -> (Vec<ArtilleryMember>, Vec<ArtilleryMember>) {
-        let mut current_members = self.to_map();
+        let mut current_members = self.members.clone();
 
         let mut changed_nodes = Vec::new();
         let mut new_nodes = Vec::new();
 
-        let my_node_id = self.mut_myself().node_id();
+        let my_node_id = self.current_node_id();
 
         for state_change in state_changes {
-            let new_member_data = state_change.member();
-            let old_member_data = current_members.entry(new_member_data.node_id());
+            let member_change = state_change.member();
 
-            if new_member_data.node_id() == my_node_id {
-                if new_member_data.state() != ArtilleryMemberState::Alive {
+            if member_change.node_id() == my_node_id {
+                if member_change.state() != ArtilleryMemberState::Alive {
                     let myself = self.reincarnate_self();
                     changed_nodes.push(myself.clone());
                 }
-            } else {
-                match old_member_data {
-                    Entry::Occupied(mut entry) => {
-                        let new_member =
-                            member::most_uptodate_member_data(new_member_data, entry.get()).clone();
-                        let new_host = new_member
-                            .remote_host()
-                            .or_else(|| entry.get().remote_host())
-                            .unwrap();
-                        let new_member = new_member.member_by_changing_host(new_host);
-
-                        if new_member.state() != entry.get().state() {
-                            entry.insert(new_member.clone());
-                            changed_nodes.push(new_member);
-                        }
+            } else if let Some(existing_member) = Self::find_member(
+                &mut current_members,
+                &member_change.node_id(),
+                member_change.remote_host().unwrap_or(*from),
+            ) {
+                let update_member =
+                    member::most_uptodate_member_data(member_change, existing_member).clone();
```

Comment:
When looking at `most_uptodate_member_data` this will not work. `update_member` will just keep the same node_id because in the common case where you kill one node and restart immediately, `member_change` and `existing_member` will still have the same state and same incarnation number and in this case, `most_uptodate_member_data` will return the `existing_member`. And now you lose the new `node_id`

Reply:
the node id update should be covered by Heartbeat(node_id) -> Ack(node_id)